proposals/NNNN-resource-constructors.md

```diff
@@ -0,0 +1,228 @@
+variable declaration processing (`Sema::ActOnUninitializedDecl`). It will
+work as if it would replace:
+
+`RWBuffer<float> A : register(u3);`
+
+with
+
+`RWBuffer<float> A(3,0,1,0);`.
+
+### Constructor for resources with implicit binding
+
+If a resource does not have an explicit binding annotation, or if it has one but
+it only specifies the virtual register space, it has _implicit binding_. The
+actual binding will be assigned later on by the compiler.
+
+The constructor for resources with implicit binding looks like this:
+
+```c++
+template <typename T> struct RWBuffer {
+  ...
+private:
+  // For resources with implicit binding
+  RWBuffer(unsigned space, int range, unsigned index) {
+    __handle = __builtin_hlsl_resource_createhandlefromimplicitbinding(__handle, space, range, index);
+  }
+  ...
+};
+```
+
+The `__handle` argument passed into the
+`__builtin_hlsl_resource_createhandlefromimplicitbinding` Clang builtin function will
+be used to infer the return type of the that function (the same way as for the
+default construtor).
+
+A call to this constructor will be created by Sema as part of uninitialized
+variable declaration processing (`Sema::ActOnUninitializedDecl`). It will
+work as if it would replace:
+
+`RWBuffer<float> A;`
+
+with
+
+`RWBuffer<float> A(0,1,0);`.
+
+Or if the resource has a space-only binding annotation, it will work as if it
+would replace:
+
+`RWBuffer<float> A : register(space13);`
+
+with
+
+`RWBuffer<float> A(13,1,0);`.
+
+### Copy constructor and assignment operator
+
+The copy constructor and/or the assignemt operator will take care of copying the
+resource handle between resource class instances.
+
+```c++
+template <typename T> struct RWBuffer {
+  ...
+public:
+  // Resources are copyable.
+  RWBuffer(RWBuffer &LHS) = default;
+
+  // Resources are assignable.
+  RWBuffer &operator=(RWBuffer &LHS) {
+    __handle = LHS.__handle;
+    return *this
+  }
+  ...
+};
+```
+
+### Dynamically-bound resources
+
+TBD
+
+### Summary
+
+```c++
+template <typename T> struct RWBuffer {
+private:
+  // resource handle
+  using handle_t = __hlsl_resource_t
+      [[hlsl::contained_type(T)]] [[hlsl::resource_class(UAV)]];
+  handle_t __handle;
+
+  // For resources with explicit binding
+  RWBuffer(unsigned register, unsigned space, int rage, unsigned index) {
+    __handle = __builtin_hlsl_resource_createhandlefrombinding(__handle, register, space, range, index);
+  }
+
+  // For resources with implicit binding
+  RWBuffer(unsigned space, int rage, unsigned index) {
+    __handle = __builtin_hlsl_resource_createhandlefromimplicitbinding(__handle, space, range, index);
+  }
+
+public:
+  // For uninitialized handles
+  RWBuffer() {
+    __handle = __builtin_hlsl_resource_createpoisonhandle(__handle);
+  }
+
+  // Resources are copyable.
+  RWBuffer(RWBuffer &LHS) = default;
+
+  // Resources are assignable.
+  RWBuffer &operator=(RWBuffer &LHS) {
+    __handle = LHS.__handle;
```

Comment:
I think __handle is the only field in RWBuffer, so I don't understand why this function can't also be `= default;`'d?